### ISSUE TITLE
internal tests launch executable on every test run

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/cargo#5965]
There is something strange in the cargo tests.
On every run, it launches a third party program (calibre) to open some documents.
To reproduce, run `cargo test workspace_open` in the cargo repo.
The behavior is caused by these tests:
```
doc::doc_workspace_open_different_library_and_package_names
doc::doc_workspace_open_binary
doc::doc_workspace_open_binary_and_library
```
This is annoying when running the test suit because I have to close 3 windows after every run.

**Comment by matthiaskrgr** [comment]
Probably caused by  cargo doc --open 
https://github.com/rust-lang/cargo/blob/9ab347d2203b924d303940df468748185a45eea1/tests/testsuite/doc.rs#L991

cc @dwijnand

**Comment by dwijnand** [comment]
I touched that line last for test suite cleanup reasons, @debris is the original author (#5024).

Perhaps this behaviour is due to the change to `--open` in #5888 by @Seeker14491?

**Comment by Seeker14491** [comment]
I see the problem.

https://github.com/rust-lang/cargo/blob/9ab347d2203b924d303940df468748185a45eea1/tests/testsuite/doc.rs#L1028-L1032

We set the `BROWSER` environment variable to `echo`, which would previously override what program `cargo doc --open` would use to open the generated docs. This behavior was removed with my changes in https://github.com/rust-lang/cargo/pull/5888; the `xdg-open` script is always used now, and completely handles finding a program. `xdg-open` does look at the `BROWSER` environment variable, but only as a fallback.

A note: the previous behavior of the `BROWSER` override was not implemented on Windows or Mac, which is why these tests all have the cfg

https://github.com/rust-lang/cargo/blob/9ab347d2203b924d303940df468748185a45eea1/tests/testsuite/doc.rs#L1007

I'm not sure what the best way to fix these tests would be. Even if the override were implemented, it really wouldn't be testing the normal operation of the `--open` flag, which is opening in the user's default browser.
